File: print-binary-tree.cpp
//Runtime: 3 ms
class Solution {
public:
    int findH(TreeNode* root)
    {
        if (!root)
            return 0;

        int a, b;
        a = findH(root->left);
        b = findH(root->right);

        return 1 + max(a, b);
    }

    void solve(vector<vector<string> > &res, TreeNode* root, int l, int r, int h)
    {
        if (!root || l > r)
            return;

        int mid = l + (r - l) / 2;

        res[h][mid] = to_string(root->val);
        solve(res, root->left, l, mid, h + 1);
        solve(res, root->right, mid + 1, r, h + 1);
    }

    vector<vector<string>> printTree(TreeNode* root) {
        int h = findH(root);
        vector<vector<string> >res(h, vector<string>(pow(2, h) - 1, ""));
        solve(res, root, 0, pow(2, h) - 1, 0);
        return res;
    }
};

File: print-binary-tree.py
# Runtime: 48 ms
class Solution(object):
    def getHeight(self, root):
        if not root:
            return 0

        return 1 + max(self.getHeight(root.left), self.getHeight(root.right))

    def solve(self, root, res, h, lo, hi):
        if not root or lo > hi:
            return

        mid = (lo + hi) / 2
        res[h][mid] = str(root.val)
        self.solve(root.left, res, h + 1, lo, mid)
        self.solve(root.right, res, h + 1, mid + 1, hi)

    def printTree(self, root):
        h = self.getHeight(root)
        res = [['' for _ in range(pow(2, h) - 1)] for _ in range(h)]
        self.solve(root, res, 0, 0, pow(2, h) - 1)
        return res
